src/Mod/Draft/draftmake/make_copy.py

```diff
@@ -61,152 +44,40 @@ def make_copy(obj, force=None, reparent=False):
-
-    elif (utils.get_type(obj) == "BSpline") or (force == "BSpline"):
-        _name = utils.get_real_name(obj.Name)
-        newobj = App.ActiveDocument.addObject(obj.TypeId, _name)
-        BSpline(newobj)
-        if App.GuiUp:
-            ViewProviderWire(newobj.ViewObject)
-
-    elif (utils.get_type(obj) == "Block") or (force == "BSpline"):
-        _name = utils.get_real_name(obj.Name)
-        newobj = App.ActiveDocument.addObject(obj.TypeId, _name)
-        Block(newobj)
-        if App.GuiUp:
-            ViewProviderDraftPart(newobj.ViewObject)
-            
-    # drawingview became obsolete with v 0.19
-    # TODO: uncomment after splitting DrawingView object from draft py
-
-    #elif (utils.get_type(obj) == "DrawingView") or (force == "DrawingView"):
-    #_name = utils.get_real_name(obj.Name)
-    #newobj = App.ActiveDocument.addObject(obj.TypeId, _name)
-    #DrawingView(newobj)
-
-    elif (utils.get_type(obj) == "Structure") or (force == "Structure"):
-        import ArchStructure
-        _name = utils.get_real_name(obj.Name)
-        newobj = App.ActiveDocument.addObject(obj.TypeId, _name)
-        ArchStructure._Structure(newobj)
-        if App.GuiUp:
-            ArchStructure._ViewProviderStructure(newobj.ViewObject)
-
-    elif (utils.get_type(obj) == "Wall") or (force == "Wall"):
-        import ArchWall
-        _name = utils.get_real_name(obj.Name)
-        newobj = App.ActiveDocument.addObject(obj.TypeId, _name)
-        ArchWall._Wall(newobj)
-        if App.GuiUp:
-            ArchWall._ViewProviderWall(newobj.ViewObject)
-
-    elif (utils.get_type(obj) == "Window") or (force == "Window"):
-        import ArchWindow
-        _name = utils.get_real_name(obj.Name)
-        newobj = App.ActiveDocument.addObject(obj.TypeId, _name)
-        ArchWindow._Window(newobj)
-        if App.GuiUp:
-            ArchWindow._ViewProviderWindow(newobj.ViewObject)
-
-    elif (utils.get_type(obj) == "Panel") or (force == "Panel"):
-        import ArchPanel
-        _name = utils.get_real_name(obj.Name)
-        newobj = App.ActiveDocument.addObject(obj.TypeId, _name)
-        ArchPanel._Panel(newobj)
-        if App.GuiUp:
-            ArchPanel._ViewProviderPanel(newobj.ViewObject)
-
-    elif (utils.get_type(obj) == "Sketch") or (force == "Sketch"):
-        _name = utils.get_real_name(obj.Name)
-        newobj = App.ActiveDocument.addObject("Sketcher::SketchObject", _name)
-        for geo in obj.Geometry:
-            newobj.addGeometry(geo)
-        for con in obj.Constraints:
-            newobj.addConstraint(con)
-
-    elif hasattr(obj, 'Shape'):
+    if simple_copy and hasattr(obj, 'Shape'):
         _name = utils.get_real_name(obj.Name)
         newobj = App.ActiveDocument.addObject("Part::Feature", _name)
         newobj.Shape = obj.Shape
+        gui_utils.format_object(newobj, obj)
+    elif not simple_copy and hasattr(obj, 'Shape'):
+        newobj = App.ActiveDocument.copyObject(obj)
```

Comment:
The code looks fine, although I'm not sure what it's supposed to do.

With `simple_copy` we want to obtain a non-parametric object? Is this right? How did this work before?

Otherwise, we want to have a "complex copy", meaning a scripted object of the same class and the same properties, is this right?

Reply:
exactly.
Before the simple copy option was chosen automatically when object was not one of the special cases handled and object had a shape.
So for example a Draft.Rectangle was fully copied, but a Part::Box was not, resulting in a simple Part::Feature.
With this implementation all objects can be fully copied or just simply copied. (thanks to realthunder Doc.copyObject)